keps/sig-apps/0012-20180522-deployment-inplace-update.md

```diff
@@ -0,0 +1,204 @@
+---
+kep-number: 12
+title: Support deployment controller in-place update
+authors:
+  - "@jian-he"
+owning-sig: sig-apps
+reviewers:
+  - @janetkuo, @kargakis, @smarterclayton
+approvers:
+  - TBD
+editor: "@jian-he"
+creation-date: "2018-05-23"
+last-updated: "2018-05-31"
+status: provisional
+---
+
+# Support deployment controller in-place update
+
+Table of Contents
+=================
+
+   * [Support deployment controller in-place update](#support-deployment-controller-in-place-update)
+   * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+      * [Goals](#goals)
+      * [Proposal](#proposal)
+         * [API Change](#api-change)
+         * [Deployment Controller](#deployment-controller)
+            * [Update](#update)
+            * [Rollback](#rollback)
+         * [Alternatives](#alternatives)
+            * [Alternative 1](#alternative-1)
+            * [Alternative 2](#alternative-2)
+            
+## Summary
+We propose a new update strategy for deployment controller to do rolling update without tearing down the pods, i.e. in-place update.
+
+
+## Motivation
+Today, Deployment controller, while doing rolling update, will teardown all the pods and then recreate all the pods.
+This approach works in most cases but has several limitations:
+* The re-created pod may be allocated onto different nodes, but certain pods depend on local state on the node. 
+
+    For example: some pods have persistent state in local storage. 
+    Especially when we do high pressure with very large scale testing, we may go through several rounds of upgrades. If the pods randomly run on any nodes, 
+    the network environment is not consistent, it becomes hard to guarantee a stable result.
+* The new pod may not be allocated due to resources taken by other pods. Preemption can mitigate this situation, but this 
+introduces an extra scheduling loop that can be avoided.
+* The IP address may be changed due to the pod is re-created, but we want fixed IP across container upgrades. 
```

Comment:
I don't think this is a reasonable property to ensure in the Deployment controller

Reply:
This probably would make more sense in StatefulSet.